Mathlib/NumberTheory/ModularForms/Basic.lean

```diff
@@ -397,3 +399,125 @@ instance (priority := 99) [CuspFormClass F Γ k] : ModularFormClass F Γ k where
   bdd_at_infty _ _ := (CuspFormClass.zero_at_infty _ _).boundedAtFilter
 
 end CuspForm
+
+namespace ModularForm
+
+section GradedRing
+
+/--cast for modular forms, which is useful for removing `heq`'s.-/
+def mcast {a b : ℤ} {Γ : Subgroup SL(2, ℤ)} (h : a = b) (f : ModularForm Γ a) : ModularForm Γ b
+    where
+  toFun := (f : ℍ → ℂ)
+  slash_action_eq' := by intro A; have := f.slash_action_eq' A; convert this; exact h.symm
+  holo' := f.holo'
+  bdd_at_infty' := by intro A; convert f.bdd_at_infty' A <;> exact h.symm
+
+theorem type_eq {a b : ℤ} (Γ : Subgroup SL(2, ℤ)) (h : a = b) :
+    ModularForm Γ a = ModularForm Γ b := by
+  induction h
+  rfl
+
+theorem cast_eq_mcast {a b : ℤ} {Γ : Subgroup SL(2, ℤ)} (h : a = b) (f : ModularForm Γ a) :
+    cast (type_eq Γ h) f = mcast h f := by
+  induction h
+  ext1
+  rfl
```

Comment:
I think you can remove the need for these if you instead add:
```lean
@[ext]
theorem gradedMonoid_eq_of_cast {Γ : Subgroup SL(2, ℤ)} {a b : GradedMonoid (ModularForm Γ)}
    (h : a.fst = b.fst) (h2 : mcast h a.snd = b.snd) : a = b := by
  obtain ⟨i, a⟩ := a
  obtain ⟨j, b⟩ := b
  cases h
  exact congr_arg _ h2
```
which is the approach I used for tensor powers

Reply:
Sorry if this is a stupid question, but I just want to check I understand. You're saying using this I can get rid of `type_eq` and `cast_eq_mcast`, but I then also need to change the `hEq` lemmas bellow to lemmas about `mcast` instead. For it work like in the tensor powers.

Reply:
Yes, that is indeed what I mean